crates/syntax/src/ast/edit_in_place.rs

```diff
@@ -257,6 +290,28 @@ impl ast::GenericParamList {
         }
     }
 
+    pub fn syntax_editor_add_generic_param(
+        &self,
+        editor: &mut SyntaxEditor,
+        new_param: ast::GenericParam,
+    ) {
+        match self.generic_params().last() {
+            Some(_) => {
+                let mut params =
+                    self.generic_params().map(|param| param.clone()).collect::<Vec<_>>();
+                params.push(new_param.into());
+                let new_param_list = make::generic_param_list(params);
+
+                editor.replace(self.syntax(), new_param_list.syntax());
+            }
+            None => {
+                let position = crate::syntax_editor::Position::after(self.l_angle_token().unwrap());
+                let new_param_list = make::generic_param_list(once(new_param.clone()));
+                editor.insert(position, new_param_list.syntax());
```

Comment:
This panics on the following assertion;

https://github.com/rust-lang/rust-analyzer/blob/a2bb0149144e0e6330e60abe4683b4e1a7ab3582/crates/syntax/src/syntax_editor.rs#L54-L55

I think that the panic happens when we are here in the following process;

1. First, we insert a new empty generic arg list because the function doesn't have one
2. And then we add a new type parameter for impl Trait thing

In 2. we don't have the new empty generic arg list made in 1. in our syntax tree yet (it will be made when the syntax edit finishes) so the position for new type parameter doesn't exist in our syntax tree yet either.
So, I think that we should directly insert a new generic arg list with new type param in 1. without separated step 2.

(Also, the current logic seems problematic beside the above because we are inserting `<T>` into `<>`, instead of `T`)

Reply:
So just for context, there’s been some [discussion about this in Zulip](https://rust-lang.zulipchat.com/#narrow/channel/185405-t-compiler.2Frust-analyzer/topic/.60SyntaxEditor.60.20replacement.20for.20Mutable.20Syntax.20trees/near/483142487).

> In 2. we don't have the new empty generic arg list made in 1. in our syntax tree yet (it will be made when the syntax edit finishes) so the position for new type parameter doesn't exist in our syntax tree yet either.

That was exactly the issue. I updated the code to add a method, `ast::Fn::syntax_editor_add_generic_param`, that handles the logic for us. Now, there’s a clear separation between the three cases in this method:
- If there is an existing generic param list and it’s not empty, we replace the entire list with a new one (existing params + the new one).
- If there’s an existing but empty generic param list, we place the new param after the left angle token.
- If there is no generic param list, we create one from scratch and place it in the right position.

This is working fine for all these cases, but the only issue left is the cursor position—it’s off in all cases.

I updated the code to use `SyntaxEditor::add_annotation()` as suggested https://github.com/rust-lang/rust-analyzer/pull/18483#discussion_r1856919754 , but I'm still facing the cursor issue. After tracing it a bit, it seems that this might be happening because `fn_.generic_param_list()` isn’t being updated. If that’s the case, how can I update it?